## Gather

In [2]:
import pandas as pd
import re as re

In [3]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [4]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [5]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [6]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [7]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [8]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [9]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [10]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [11]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [12]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [13]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [14]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [15]:
patients.sample(5)

patient_id assigned_sex given_name       surname             address  \
297         298         male      Bodor        Csonka  323 Platinum Drive   
423         424       female     Simone  Baumgaertner    510 Berry Street   
258         259         male       Abel       Yonatan    2621 Koontz Lane   
30           31       female     Janina      Bazarova      3259 Roy Alley   
374         375         male       Ingi     Þráinsson  707 Goodwin Avenue   

                 city       state  zip_code        country  \
297           Crafton          PA   15205.0  United States   
423  Colorado Springs          CO   80904.0  United States   
258           Burbank  California   91502.0  United States   
30          Englewood          CO   80112.0  United States   
374            Colfax          WA   99111.0  United States   

                                        contact  birthdate  weight  height  \
297      724-759-0310CsonkaBodor@jourrapide.com  7/18/1993   157.3      67   
423  SimoneBaumgaertner@teleworm.us719-661-6624  5/14/1931   125.2      59   
258       AbelYonatan@teleworm.us1 818 841 7660  4/29/1952   137.9      66   
30     303-910-2058JaninaBazarova@superrito.com  6/29/1964   156.9      65   
374         509-397-8510Ingirainsson@dayrep.com   2/5/1986   210.8      69   

      bmi  
297  24.6  
423  25.3  
258  22.3  
30   26.1  
374  31.1

In [16]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Lương          2
Souza          2
Batukayev      2
Nilsen         2
Woźniak        2
Berg           2
Hueber         2
Kowalczyk      2
Tucker         2
Silva          2
Liễu           2
Correia        2
Bùi            2
Ogochukwu      2
Tạ             2
Grímsdóttir    2
Parker         2
Collins        2
Schiavone      2
Dratchev       2
Aranda         2
Johnson        2
Kadyrov        2
Lâm            2
Cabrera        2
Gersten        2
Cindrić        2
              ..
Su             1
Montez         1
Crawford       1
Fomin          1
Chin           1
Chang          1
Endrizzi       1
Seppälä        1
Bakos          1
Sandgreen      1
Lansell        1
Archer         1
Naifeh         1
Ferrari        1
Maríasson      1
Martinsson     1
Holm           1
Csorba         1
Wellish        1
Phạm           1
Onwughara      1
Teichelmann    1
Sjöberg        1
Nicholls       1
Holden         1
Nnonso         1
She            1
MacDonald     

In [17]:
patients.address.value_counts()

123 Main Street               6
2476 Fulton Street            2
2778 North Avenue             2
648 Old Dear Lane             2
1092 Deans Lane               1
3343 Jefferson Street         1
4646 Highland View Drive      1
4386 Camden Street            1
2917 Golden Ridge Road        1
510 Berry Street              1
3402 Kildeer Drive            1
3488 Clair Street             1
4271 Cherry Ridge Drive       1
1233 Liberty Avenue           1
321 Briercliff Road           1
3434 Holt Street              1
1840 Millbrook Road           1
1343 Clair Street             1
3920 Braxton Street           1
3084 Blue Spruce Lane         1
1753 Sherman Street           1
3259 Roy Alley                1
2166 Tipple Road              1
1033 Wayside Lane             1
2389 Rubaiyat Road            1
1092 Farm Meadow Drive        1
4982 Wood Street              1
2595 Feathers Hooves Drive    1
989 Wayback Lane              1
3300 Woodridge Lane           1
                             ..
3216 Lod

In [18]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [19]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [20]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [21]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [22]:
sum(treatments.auralin.isnull())

0

In [23]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [24]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
*Your definition here. Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*

In [25]:
df = pd.read_csv('treatments_cut.csv')

In [26]:
df.shape

(70, 7)

##### Code

In [27]:
# Your cleaning code here
#here, i try to join my missing data to the original dataset

df_join = pd.concat([treatments_clean, df], join = 'inner', ignore_index = True, verify_integrity=True)


##### Test

In [28]:
# Your testing code here
#now i will check the info of my newly joined table
df_join.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    213 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


In [29]:
df_join.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
*Your definition here.*
- here, i will be calculating the hba1c change and refilling the change column with my results
- thus, accounting for the missing and innacurate data

##### Code

In [30]:
# Your cleaning code here
#i will subtract the start level from the end level
#here, i will load my data

df_join.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [31]:
#here, i will do the calculation
df_join['hba1c_change'] = df_join['hba1c_start'] - df_join['hba1c_end']

##### Test

In [32]:
# Your testing code here
#this code checks for null values in the hba1c_change column
df_join.hba1c_change.isnull().sum()

0

In [33]:
#for inaccurate values, this code confirms that all calculations is correct

df_join.sample(5)

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
51   kristoffer  martinsen  29u - 37u          -         9.18       8.64   
7         eddie     archer  31u - 38u          -         7.89       7.55   
197    angelico   marchesi          -  58u - 58u         7.62       7.18   
311       galit    casárez  27u - 37u          -         7.91       7.56   
256       ilija     horvat  42u - 50u          -         7.77       7.38   

     hba1c_change  
51           0.54  
7            0.34  
197          0.44  
311          0.35  
256          0.39

In [34]:
treatments_clean = df_join

In [35]:
treatments_clean.sample(4)

given_name  surname    auralin    novodra  hba1c_start  hba1c_end  \
245           wu     sung          -  47u - 48u         7.61       7.12   
22   brancaleone    russo  53u - 60u          -         8.61       8.18   
106        sofia  karlsen          -  35u - 29u         7.62       7.15   
54          oles  zhdanov  54u - 67u          -         7.52       7.11   

     hba1c_change  
245          0.49  
22           0.43  
106          0.47  
54           0.41

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

##### Code

In [36]:
# Your cleaning code here
#displaying the contact column to be worked on
patients_clean.contact

0                   951-719-9170ZoeWellish@superrito.com
1                  PamelaSHill@cuvox.de+1 (217) 569-3204
2                       402-363-6804JaeMDebord@gustr.com
3             PhanBaLiem@jourrapide.com+1 (732) 636-8246
4                        334-515-7487TimNeudorf@cuvox.de
5               386-334-5237RafaelCardosoCosta@gustr.com
6                      775-533-5933MaryBAdams@einrot.com
7                   XiuxiuChang@einrot.com1 408 778 3236
8                816-265-9578DavidGustafsson@armyspy.com
9          SophieCabreraIbarra@teleworm.us1 718 795 9124
10                906-478-8949SandyGunnarsson@dayrep.com
11            Abdul-NurMummarIsa@rhyta.com1 931 207 0839
12             OmeokachieIbeamaka@einrot.com434-509-2614
13            AnenechiChidi@armyspy.com+1 (205) 417-8095
14                     AsiaWozniak@rhyta.com918-712-3469
15                       276-225-1955SrenFLund@gustr.com
16                LieuThiThuTam@dayrep.com1 559 765 7836
17              RoxanneAndreyev

In [37]:
#tried to create a pattern of regular expresions
email_pattern = r'([^0-9-._/|,\s][a-zA-Z0-9_.+-]+@[a-zA-Z-]+\.[a-zA-Z-.]+)'

phone_pattern = r'(((\+?\d+)?[\s.-]?)?\(?[2-9]?(?!11)\d{2}\)?[\s.-]?\d{3}[\s.-]?\d{4})'

In [38]:
#this code extracts the email from the contact column
patients_clean['Email'] = patients_clean.contact.str.extract(email_pattern, flags = 0)


In [39]:
#here, i extracted the phone numbers from the contact column
#the p1 and p2 were created temporarily to store for irrelevant values returned by the regex pattern
patients_clean[['phone_number', 'p1', 'p2']] = patients_clean.contact.str.extract(phone_pattern, flags = 0)

In [40]:
#next i will drop irrellevant columns created for the sake of analysis
patients_clean.drop(columns = ['p1', 'p2', 'contact'], inplace = True)

##### Test

In [41]:
# Your testing code here
#this code shows the columns created 
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   
2               York    Nebraska   68467.0  United States  2/19/1980   177.8   
3         Woodbridge          NJ    7095.0  United States  7/26/1951   220.9   
4             Dothan          AL   36303.0  United States  2/18/1928   192.3   

   height   bmi                      Email       phone_number  
0      66  19.6   ZoeWellish@superrito.com       951-719-9170  
1      66  19.2       PamelaSHill@cuvox.de  +1 (217) 569-3204  
2      71  24.8       JaeMDebord@gustr.com       402-363-6804  
3      70  31.7  PhanBaLiem@jourrapide.com  +1 (732) 636-8246  
4      27  26.1        TimNeudorf@cuvox.de       334-515-7487

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

In [42]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

##### Code

In [43]:
# Your cleaning code here
#here, i try to write a code that transforms the column names, auralin and novodra into value in a new column 'treatments' 
#and then set it values
treatments_clean = pd.melt(treatments_clean, id_vars=["given_name", 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], 
                 var_name="Treatment", value_name="Dosage")

In [44]:
#this code shows that the above task is confirmed
treatments_clean.head(5)

given_name     surname  hba1c_start  hba1c_end  hba1c_change Treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1     elliot  richardson         7.56       7.09          0.47   auralin   
2   yukitaka    takenaka         7.68       7.25          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
4     alissa      montez         7.78       7.46          0.32   auralin   

      Dosage  
0  41u - 48u  
1          -  
2          -  
3  33u - 36u  
4          -

In [45]:
#this code is used to drop rows where there is no dosage values, represented as '-'
treatments_clean.drop(treatments_clean[treatments_clean.Dosage.str.contains(r'^[\W-]+$')].index, inplace = True)

#this code confirms the above results
treatments_clean.shape

(350, 7)

In [46]:
#here, i split the dosage column to get the start and end dose respectively
treatments_clean[['start_dose', 'end_dose']] = treatments_clean.Dosage.str.split("-", n=1, expand=True)

In [47]:
#this code confirms the above operation
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change Treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

      Dosage start_dose end_dose  
0  41u - 48u       41u       48u  
3  33u - 36u       33u       36u  
6  37u - 42u       37u       42u  
7  31u - 38u       31u       38u  
9  30u - 36u       30u       36u

In [48]:
#this line of code removes the non-number characters in the dosage column
treatments_clean['start_dose'] = treatments_clean['start_dose'].str.strip('u ')
treatments_clean['end_dose'] = treatments_clean['end_dose'].str.strip('u ')

#this line of code drops the irrelevant columns
treatments_clean.drop(columns = ['Dosage'], inplace = True)

#this line of code aims to reset index of the dataframe
treatments_clean = treatments_clean.reset_index(drop = True)

##### Test

In [49]:
# Your testing code here
#this code checks that all operations performed above are effected
treatments_clean.sample(5)

given_name      surname  hba1c_start  hba1c_end  hba1c_change Treatment  \
11   brancaleone        russo         8.61       8.18          0.43   auralin   
67         ruman      bisliev         7.72       7.39          0.33   auralin   
322       edelma  villalpando         7.99       7.56          0.43   novodra   
188       suhaim        rahal         7.94       7.50          0.44   novodra   
320         jane      citizen         7.98       7.60          0.38   novodra   

    start_dose end_dose  
11          53       60  
67          46       53  
322         24       26  
188         49       47  
320         37       38

In [50]:
#this code also helps to check that the number of samples still remains 350
treatments_clean[treatments_clean['given_name'].duplicated()]

given_name     surname  hba1c_start  hba1c_end  hba1c_change Treatment  \
62      joseph         day         7.70       7.19          0.51   auralin   
106  alexander      hueber         7.71       7.35          0.36   auralin   
128    rebecca    jephcott         7.96       7.57          0.39   auralin   
167     joseph      tucker         7.67       7.30          0.37   auralin   
168     robert      maslov         7.65       7.35          0.30   auralin   
244      david  gustafsson         7.72       7.28          0.44   novodra   
246      søren    sørensen         7.93       7.59          0.34   novodra   
254    mustafa   lindström         7.92       7.60          0.32   novodra   
293       anja      hueber         8.82       8.36          0.46   novodra   
308      david    beauvais         7.87       7.47          0.40   novodra   
313     yasmin      araujo         7.82       7.36          0.46   novodra   
333      sofia   hermansen         8.90       8.57          0.33   novodra   

    start_dose end_dose  
62          29       36  
106         35       38  
128         53       63  
167         48       56  
168         55       68  
244         33       34  
246         41       39  
254         35       33  
293         31       32  
308         26       23  
313         51       54  
333         34       34

#### Adverse reaction should be part of the `treatments` table

##### Define
*Your definition here: here, i am to merge my adverse reaction samples to their respective sample on the treatments table

Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

In [51]:
#checks the number of samples to be merged
adverse_reactions_clean.shape

(34, 3)

##### Code

In [52]:
# Your cleaning code here
#here i want to try to merge the adverse reactions table to my treatments table on patients with given names and surnames

treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on=['given_name', 'surname'], how='left')

##### Test

In [53]:
# Your testing code here
#to confirm the above operation
treatments_clean.head(10)

given_name       surname  hba1c_start  hba1c_end  hba1c_change Treatment  \
0   veronika      jindrová         7.63       7.20          0.43   auralin   
1       skye    gormanston         7.97       7.62          0.35   auralin   
2     sophia        haugen         7.65       7.27          0.38   auralin   
3      eddie        archer         7.89       7.55          0.34   auralin   
4       asia       woźniak         7.76       7.37          0.39   auralin   
5     joseph           day         7.70       7.19          0.51   auralin   
6    roxanne     andreyeva         9.54       9.14          0.40   auralin   
7     simone  baumgaertner         7.74       7.30          0.44   auralin   
8       enco        žibrik         7.78       7.34          0.44   auralin   
9    camilla      zaitseva         7.53       7.13          0.40   auralin   

  start_dose end_dose adverse_reaction  
0         41       48              NaN  
1         33       36              NaN  
2         37       42              NaN  
3         31       38              NaN  
4         30       36              NaN  
5         29       36     hypoglycemia  
6         29       38              NaN  
7         27       37              NaN  
8         55       68              NaN  
9         28       37              NaN

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

##### Code

In [54]:
# Your cleaning code here
#here i will try to create a dataframe to store my required columns from patients
names_rep = patients_clean[['given_name', 'surname', 'patient_id']]

In [55]:
#to turn all entries of names to lower letters for consistency
names_rep.given_name = names_rep.given_name.str.lower()
names_rep.surname = names_rep.surname.str.lower()

/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [56]:
#merge the new index columns to the treatments_clean so as to eliminate names in the table and replace with patients_id
treatments_clean = pd.merge(treatments_clean, names_rep, on = ['given_name', 'surname'])

In [57]:
#dropping irrelevant columns
treatments_clean.drop(columns = ['given_name', 'surname'], inplace = True)

##### Test

In [58]:
# Your testing code here
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change Treatment start_dose end_dose  \
0         7.63       7.20          0.43   auralin         41       48   
1         7.97       7.62          0.35   auralin         33       36   
2         7.65       7.27          0.38   auralin         37       42   
3         7.89       7.55          0.34   auralin         31       38   
4         7.76       7.37          0.39   auralin         30       36   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
*Your definition here: i will change the data type of the zip code and also remove all '.0' with string slicing after which i will fill the zipcode to make sure all length id 5 characters length.*

*Hint: see the "Data Cleaning Process" page.*

##### Code

In [59]:
# Your cleaning code here
#this line of code converts data type to string, removes the decimal points and makes sure all zipcodes are 5 characters long
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar = '0')



In [60]:
import numpy as np
#this line of code replaces all '0000n' with NAN 
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

##### Test

In [61]:
# Your testing code here
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ    07095  United States  7/26/1951   220.9   
4             Dothan          AL    36303  United States  2/18/1928   192.3   

   height   bmi                      Email       phone_number  
0      66  19.6   ZoeWellish@superrito.com       951-719-9170  
1      66  19.2       PamelaSHill@cuvox.de  +1 (217) 569-3204  
2      71  24.8       JaeMDebord@gustr.com       402-363-6804  
3      70  31.7  PhanBaLiem@jourrapide.com  +1 (732) 636-8246  
4      27  26.1        TimNeudorf@cuvox.de       334-515-7487

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Your definition here: i want to replace the value of Tim Neudorf height*

##### Code

In [62]:
# Your cleaning code here
#here, i show the exact height of tim and discovered he is the only outlier with 27
patients_clean[patients_clean.given_name == 'Tim']


patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      27  26.1   

                 Email  phone_number  
4  TimNeudorf@cuvox.de  334-515-7487

In [63]:
#this code replaces the only outlier in our height column
patients_clean['height'] = patients_clean['height'].replace([27],72)


##### Test

In [64]:
# Your testing code here
patients_clean[patients_clean.given_name == 'Tim']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

                 Email  phone_number  
4  TimNeudorf@cuvox.de  334-515-7487

- **Now, the above output confirms that my change has been effected and in place**

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

##### Code

In [65]:
# Your cleaning code here
patients_clean.state.unique()

array(['California', 'Illinois', 'Nebraska', 'NJ', 'AL', 'Florida', 'NV',
       'CA', 'MO', 'New York', 'MI', 'TN', 'VA', 'OK', 'GA', 'MT', 'MA',
       'NY', 'NM', 'IL', 'LA', 'PA', 'CO', 'ME', 'WI', 'SD', 'MN', 'FL',
       'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD', 'AZ', 'TX',
       'NE', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH', 'OR',
       nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

In [66]:
#this is a dictionary containg full names of abreviations in the state column
us_state_abbrev = {
            'AL': 'Alabama',
            'AK': 'Alaska',
            'AZ': 'Arizona',
            'AR': 'Arkansas',
            'CA': 'California',
            'CO': 'Colorado',
            'CT': 'Connecticut',
            'DE': 'Delaware',
            'FL': 'Florida',
            'GA': 'Georgia',
            'HI': 'Hawaii',
            'ID': 'Idaho',
            'IL': 'Illinois',
            'IN': 'Indiana',
            'IA': 'Iowa',
            'KS': 'Kansas',
            'KY': 'Kentucky',
            'LA': 'Louisiana',
            'ME': 'Maine',
            'MD': 'Maryland',
            'MA': 'Massachusetts',
            'MI': 'Michigan',
            'MN': 'Minnesota',
            'MS': 'Mississippi',
            'MO': 'Missouri',
            'MT': 'Montana',
            'NE': 'Nebraska',
            'NV': 'Nevada',
            'NH': 'New Hampshire',
            'NJ': 'New Jersey',
            'NM': 'New Mexico',
            'NY': 'New York',
            'NC': 'North Carolina',
            'ND': 'North Dakota',
            'OH': 'Ohio',
            'OK': 'Oklahoma',
            'OR': 'Oregon',
            'PA': 'Pennsylvania',
            'RI': 'Rhode Island',
            'SC': 'South Carolina',
            'SD': 'South Dakota',
            'TN': 'Tennessee',
            'TX': 'Texas',
            'UT': 'Utah',
            'VT': 'Vermont',
            'VA': 'Virginia',
            'WA': 'Washington',
            'WV': 'West Virginia',
            'WI': 'Wisconsin',
            'WY': 'Wyoming',
            'DC': 'District of Columbia',
            'MP': 'Northern Mariana Islands',
            'PW': 'Palau',
            'PR': 'Puerto Rico',
            'VI': 'Virgin Islands',
            'AA': 'Armed Forces Americas (Except Canada)',
            'AE': 'Armed Forces Africa/Canada/Europe/Middle East',
            'AP': 'Armed Forces Pacific'
        }

In [67]:
#here, this line of code maps the dictionary to the state column and also fill values not found in the dict with previous values
patients_clean.state = patients_clean.state.map(us_state_abbrev).fillna(patients_clean['state'])

##### Test

In [68]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge  New Jersey    07095  United States  7/26/1951   220.9   
4             Dothan     Alabama    36303  United States  2/18/1928   192.3   

   height   bmi                      Email       phone_number  
0      66  19.6   ZoeWellish@superrito.com       951-719-9170  
1      66  19.2       PamelaSHill@cuvox.de  +1 (217) 569-3204  
2      71  24.8       JaeMDebord@gustr.com       402-363-6804  
3      70  31.7  PhanBaLiem@jourrapide.com  +1 (732) 636-8246  
4      72  26.1        TimNeudorf@cuvox.de       334-515-7487

In [69]:
#confirms that samples quantity not tampered with
patients_clean.shape

(503, 15)

#### Dsvid Gustafsson

##### Define
*Your definition here: i will correct the typographical error noted in the name above*

##### Code

In [70]:
#this line of code replaces the typographical error with the correct spelliing
patients_clean.given_name = patients_clean.given_name.replace(['Dsvid'], 'David')

##### Test

In [71]:
# Your testing code here
#this code should return no value since the error has been corrected
patients_clean[patients_clean.given_name == 'Dsvid']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, Email, phone_number]
Index: []

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

##### Code

In [72]:
# Your cleaning code here
#this line of code converts assigned_sex, and state column to  category datatypes
patients_clean = patients_clean.astype({'assigned_sex': 'category','state': 'category'})




In [73]:
#this code converts birthdate column to datetime data type
patients_clean['birthdate'] = pd.to_datetime(patients_clean['birthdate'])

##### Test

In [74]:
# Your testing code here
patients_clean.dtypes

patient_id               int64
assigned_sex          category
given_name              object
surname                 object
address                 object
city                    object
state                 category
zip_code                object
country                 object
birthdate       datetime64[ns]
weight                 float64
height                   int64
bmi                    float64
Email                   object
phone_number            object
dtype: object

#### Multiple phone number formats

##### Define
*Your definition here. Hint:remove all non digit characters. helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

##### Code

In [75]:
# Your cleaning code here
#this line of code replaces all non didgit characters with nothing and then pads the remaining with a 1 to have country code

patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [76]:
# Your testing code here
patients_clean.phone_number.head()

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.since  this data is non revoverable, i will be removing it from my dataset*

##### Code

In [78]:
# Your cleaning code here
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [80]:
# Your testing code here
#should be empty
patients_clean[patients_clean.surname == 'Doe']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, Email, phone_number]
Index: []

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Your definition here: it was discovered that some records are that of nicknames, hence, i eill be dropping the records belonging to nicknames, in order to conform with the records on treatments table*

##### Code

In [81]:
#this code filters out the first occurence of duplicate values , which happens to be that of the reaal names and not nicknames
# Your cleaning code here
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [82]:
# Your testing code here
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city     state zip_code        country  birthdate  weight  height  \
24   Port Jervis  New York    12771  United States 1985-08-01   155.8      67   
432      Houston     Texas    77020  United States 1962-11-25   185.2      67   

      bmi                         Email phone_number  
24   24.4     JakobCJakobsen@einrot.com  18458587707  
432  29.0  KarenJakobsen@jourrapide.com  19792030438

In [84]:
#there should be no nick
patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

       state zip_code        country  birthdate  weight  height   bmi  \
97  Nebraska    68324  United States 1954-05-03   138.2      71  19.3   

                       Email phone_number  
97  PatrickGersten@rhyta.com  14028484923

In [83]:
#there should be nickname, sandy
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

             state zip_code        country  birthdate  weight  height   bmi  \
131  West Virginia    25962  United States 1960-10-23   206.1      64  35.4   
426        Florida    33179  United States 1992-09-02   186.6      69  27.6   

                          Email phone_number  
131    SandraCTaylor@dayrep.com  13044382648  
426  RogelioJTaylor@teleworm.us  13054346299

#### kgs instead of lbs for Zaitseva weight

##### Define
*Your definition here: it was discovered that Zaitseva's weight was miscalculated and recorded in kg instead of pounds.This record will be recalculated and replaced*

##### Code

In [89]:
# Your cleaning code here
patients_clean[patients_clean.surname == 'Zaitseva']

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate      weight  height  \
210  Wooster  Ohio    44691  United States 1938-11-26  107.585456      63   

      bmi                          Email phone_number  
210  19.1  CamillaZaitseva@superrito.com  13302022145

In [88]:
patients_clean['weight'] = patients_clean['weight'].replace([48.8],48.8 * 2.20462)


##### Test

In [91]:
# Your testing code here
#notice the change in weight for Zaitseva
patients_clean[patients_clean.surname == 'Zaitseva']


patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate      weight  height  \
210  Wooster  Ohio    44691  United States 1938-11-26  107.585456      63   

      bmi                          Email phone_number  
210  19.1  CamillaZaitseva@superrito.com  13302022145